### PR TITLE
Create SQLAchemy models

### PR DESCRIPTION
**Pull request by briana-mcn**
@TylerHendrickson here are the SQLAlchemy models for the program. I ran into a small problem with understanding the the Foreign Key relationships. Also, should I remove the remote `venv/` from my program? I found this line of commands but wasn't sure if I should proceed with using it.
```
git rm -r --cached some-directory
git commit -m 'Remove the now ignored directory "some-directory"'
git push origin master
```

### CODE REVIEW

**Comment by TylerHendrickson** [comment]
@briana-mcn Those commands you listed seem correct to me (although you should push to this branch, not master).

File: rock_paper_scissors/api/models.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+from sqlalchemy import Enum
+from sqlalchemy import ForeignKey
+from sqlalchemy.ext.declarative import as_declarative
+from sqlalchemy.orm import relationship
+
+
+@as_declarative()
+class ModelBase(object):
+    def __tablename__(self):
+        return self.__class__.__name__.lower()
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: You will actually want to avoid these types of shortcuts for auto-naming tables. Currently all of your models have single-word names, but that might not always be the case, and then you'll have a table for `MultiWordModel` called `multiwordmodel` instead of `multi_word_model` (which is how it _should_ be named).

You'll also see these types of shenanigans occurring with pluralizations. Some people have a crazy notion that the `Dog` model's table should be `dogs` and the `Cat` model's table should be `cats`. However, this type of convention breaks down when you have a `Fish` model- the auto-generated name for the table would be `fishs`, even though `fish` (or [perhaps](http://www.quickanddirtytips.com/education/grammar/fish-or-fishes) `fishes`) would be the obviously correct choice.

TL;DR Use `__tablename__ = 'game'` and other such explicit definitions for table names in your code. And if the above caveats aren't enough to convince you, remember that [explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/#the-zen-of-python).

File: rock_paper_scissors/api/models.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class Game(ModelBase):
+    id = Column(Integer, primary_key=True)
+    player_1_id = Column(Integer, ForeignKey('user.id'))
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    round_id = Column(Integer, ForeignKey('round.id'))
+    rounds_choice = Column(Integer)
+    rounds_decrement = Column(Integer)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: What is this column representing?

File: rock_paper_scissors/api/models.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    current_round = Column(Integer)
+    winner = Column(String, ForeignKey('user.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: You'll want to call this `winner_id`, for reasons that will become more apparent with other parts of this PR.

This would also be an `Integer` column, since the FK `user.id` is an int.

File: rock_paper_scissors/api/models.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    round = relationship('Round', back_populates='game')
+    user = relationship('User', back_populates='game', uselist=False)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: So, a `relationship` in SQLAlchemy is simply a shortcut provided to you that essentially replaces the need to do this:
```python
@property
def round(self):
    return Round.query.get(self.round_id)

@round.setter
def round(self, round_value):
    self.round_id = round_value.id
```
(There are some other benefits to using SQLAlchemy relationships, but this is the basic idea.)

In this instance, `user` is a fairly ambiguous relationship because there are 3 different FKs linking `Game` to `User`. A generic `user` relationship isn't very useful; instead, you want to provide shortcuts to the objects represented by your foreign key columns on `Game`, specifically:
```python
player_1 = relationship('User', foreign_keys=[player_1_id])
player_2 = relationship('User', foreign_keys=[player_2_id])
winner = relationship('User', foreign_keys=[winner_id])
```
A few things to note:
- The `foreign_keys` argument to `relationship` is not always necessary, but in this case it is because otherwise the exact relevant foreign key is [ambiguous](http://docs.sqlalchemy.org/en/latest/orm/join_conditions.html#handling-multiple-join-paths) since, as stated above, there are 3 different FK columns on `Game` that point to `User`. Without the argument, SQLAlchemy won't know which foreign key column to use when querying for the `User` object.
- You'll notice that I did not use `backref` or `back_populates` here. This is intentional, because explicit is better than implicit, and it's better to see explicitly-defined relationships on `User` pointing to `Game` instead of allowing SQLAlchemy to (invisibly) generate it for you.

File: rock_paper_scissors/api/models.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+class Round(ModelBase):
+    id = Column(Integer, primary_key=True)
+    player_1_choice = Column(String, ForeignKey('choice.id'))
+    player_2_choice = Column(String, ForeignKey('choice.id'))
+    winner = Column(String, ForeignKey('user.score'))
+
+    game = relationship('Game', back_populates='round')
+    user = relationship('User', back_populate='user')
+    choice = relationship('Choice', back_populate='round')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Same here- you will need two relationships pointing to `Choice`: one for `player_1_choice` and another for `player_2_choice`, each based on their respective FK columns.

File: rock_paper_scissors/api/models.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    current_round = Column(Integer)
+    winner = Column(String, ForeignKey('user.id'))
+
+    round = relationship('Round', back_populates='game')
+    user = relationship('User', back_populates='game', uselist=False)
+
+
+class Round(ModelBase):
+    id = Column(Integer, primary_key=True)
+    player_1_choice = Column(String, ForeignKey('choice.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This should be renamed to `player_1_choice_id`

File: rock_paper_scissors/api/models.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    player_2_choice = Column(String, ForeignKey('choice.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This should be renamed to `player_2_choice_id`

File: rock_paper_scissors/api/models.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    winner = Column(String, ForeignKey('user.score'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This should be renamed to `winner_id` and also needs to be an `Integer` FK column to `user.id`

File: rock_paper_scissors/api/models.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    game = relationship('Game', back_populates='round')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This needs a corresponding `game_id` column on the `Round` class that is a FK to `game.id`

File: rock_paper_scissors/api/models.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    user = relationship('User', back_populate='user')
+    choice = relationship('Choice', back_populate='round')
+
+
+class User(ModelBase):
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+    score = Column(Integer)
+    choices = Column(String, ForeignKey('choice.move'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This would be a FK to `choice.player_id`

File: rock_paper_scissors/api/models.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    game = relationship('Game', back_populates='round')
+    user = relationship('User', back_populate='user')
+    choice = relationship('Choice', back_populate='round')
+
+
+class User(ModelBase):
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+    score = Column(Integer)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: One question I have on this is whether it's meant to represent an all-time/lifetime score. Or maybe the better question: are `User`s intended to be reusable?

File: rock_paper_scissors/api/models.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    choices = Column(String, ForeignKey('choice.move'))
+
+    game = relationship('Game', back_populates='user')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This would need to be a one-to-many relationship (one user, many games). So it should probably be named `games` (plural), and will also need a way specify foreign keys:
```python
games = relationship('Game', foreign_keys=['Game.player_1_id', 'Game.player_2_id'])
```

Disclaimer: I _believe_ that specifying multiple values in the `foreign_keys` list specifies "this _or_ that", but I could be mistaken.

File: rock_paper_scissors/api/models.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    choice = relationship('Choice', back_populates='user')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Similarly, this would be `choices`

File: rock_paper_scissors/api/models.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    round = relationship('Round', back_populate='user')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This relationship should be removed- it doesn't exactly apply here because there's no direct relationship between `User` and `Round`.

Generally speaking, relationships should only be created when at least one model contains a FK to another model (or with a FK to an association table, in the case of many-to-many relationships).

File: rock_paper_scissors/api/models.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class Choice(ModelBase):
+    id = Column(Integer, primary_key=True)
+    rounds_id = Column(Integer, ForeignKey('rounds.id'))
+    player = Column(Integer, ForeignKey('user.id'))
+    move = Column('moves', Enum('Paper', 'Rock', 'Scissors'))
+
+    user = relationship('User', back_populates='choice')
+    round = relationship('Round', back_populate='choice')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Make sure you have a trailing newline at the end of your file :)

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@briana-mcn This is off to a good start. Most of the reason for there being so many comments is that many are duplicate occurrences of the same topic.

Let me know if you want to go over any of this in person.

File: rock_paper_scissors/api/models.py (hunk 1 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I thought my shortcut was being clever but your comments make sense to apply explicitly :)
Just to confirm, we don't want to use plural table names, although the plural convention for table naming exists. I scraped a little into StackOverflow of table naming conventions and it seems like there is a distinct side for or against pluralizing table names

File: rock_paper_scissors/api/models.py (hunk 2 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I thought it would be used as the value for the number of rounds left after each valid round play.  Tbh, I'm not sure how it would actually be implemented if we only have one game object created for the entire game even when rounds are played. The relationship of `Round` to `Game` is unclear to me.

File: rock_paper_scissors/api/models.py (hunk 6 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I also changed the value from`String` to `Integer`

File: rock_paper_scissors/api/models.py (hunk 7 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I also changed the value from`String` to `Integer`

File: rock_paper_scissors/api/models.py (hunk 1 above)
Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: In my opinion: Do not pluralize table names. It's a bad convention that breaks down and offers no practical benefit. :)

File: rock_paper_scissors/api/models.py (hunk 2 above)
Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: A single "Game" has (contains) many "Rounds". Each Round must point to a game explicitly via a FK column to `game.id`.

From the perspective of "Game to Rounds", this is a "One to Many" relationship.

File: rock_paper_scissors/api/models.py (hunk 9 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: Got it. I wasn't sure about this one, whether `Rounds` would have a relationship with `Game`

File: rock_paper_scissors/api/models.py (hunk 4 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: Thanks the explanation, that clears up the understanding about relationships a bit more.

I have a few questions that I've been thinking about, basically I think my lack of understanding with the  `Game` model in its' relation to the others models could be addressed by understanding:

1. When you are building models, do you think about the relations of models to models such as one-to-many, many-to-one, one-to-one, many-to-many when you are designing? 
2. When you use a foreign key in a model, is the entire instance/object that was referenced with the foreign key also accessible to that model with the foreign key?
i.e if `Game` has `User.id` reference, can we access the other values (`User.name, User.score, User.choices`) for `User.id`?
```
class Game(ModelBase):
    ...
    player_1_id = Column(Integer, ForeignKey('user.id'))
```
```
class User(ModelBase):
    ...
    id = Column(Integer, primary_key=True)
    name = Column(String)
    score = Column(Integer)
    choices = Column(String, ForeignKey('choice.move'))

    game = relationship('Game', back_populates='user')
    choice = relationship('Choice', back_populates='user')
    round = relationship('Round', back_populate='user')
```

3. If a bi-directional relationship is needed to be established between 2 tables, how do you establish the relationship without using `back_populate`/ `back_ref`?
4. Why is it that you use all the attributes referenced for the foreign_keys as a list? ie `foreign_keys=[referenced attribute]`?

 I was reading the documentation on foreign_keys arguments for relationships and came across [this](http://docs.sqlalchemy.org/en/latest/orm/join_conditions.html#handling-multiple-join-paths)

```
billing_address = relationship("Address", foreign_keys="[Customer.billing_address_id]")

billing_address = relationship("Address", foreign_keys="Customer.billing_address_id")
```
Is the string version for only one value reference? If it is, can we still wrap it in a list?

File: rock_paper_scissors/api/models.py (hunk 11 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I'm not sure about this. After putting more thought into it, I don't think User should be reusable but I don't know how a score would be kept (if it even needs to be) to be able to determine a winner.

File: rock_paper_scissors/api/models.py (hunk 4 above)
Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: 1. To be honest, I _don't_ really think about it that much, at least not as an up-front design concept. To be clear, these terms describe a particular directional relationship- if `Game` has many `Rounds`, then "Game to Round" is a "One to Many" relationship. Conversely, "Round to Game" is a "Many to One" relationship. It's worth also noting that there are more complex relationship scenarios that don't really fit any of these terms.
2. It's important to remember that foreign keys exist within the database and essentially mean that "any value in this FK column actually is representative of a value that lives in a different column (of another table)". Most (all?) relational databases enforce that if you have a `dog.breed_id` column that is a foreign key to `breed.id`, the database will reject any value added to `dog.breed_id` unless that value also exists in `breed.id`. 
So if I have this `dog` table:

| id | name   | breed_id |
|----|-----------|-------------|
| 1  | Fido     | 3             |
| 2  | Max     | 1             |
| 3  | Rover  | 3             |
| 4  | Lassie | 2             |

And this `breed` table:

| id | type              |
|----|------------------|
| 1  | poodle          |
| 2  | border collie |
| 3  | dalmatian     |

The database will prevent me from adding a row like this to the `dog` table:

| id | name        | breed_id |
|----|---------------|--------------|
| 5 | Rin Tin Tin | 10           |

because the `breed.id` column contains no value of `10`.

Now, about SQLAlchemy relationships. This is a helpful utility provided by SQLAlchemy that will query a related object for you, so you don't need to refer to the FK all the time. It comes in handy, but you should remember that they're _not_ required. Following the above example, those could be represented by the following SQLA models:
```python
class Breed(Model):
    __tablename__ = 'breed'
    id = Column(Integer, primary_key=True)
    type = Column(Text, nullable=False, unique=True)

    # We are adding this because it's convenient, 
    # not because it's necessary:
    dogs = relationship('Dog', foreign_keys='Dog.breed_id')

    def __repr__(self):
        return '<Breed {}: {}>'.format(self.id, self.type)

class Dog(Model):
    __tablename__ = 'dog'
    id = Column(Integer, primary_key=True)
    name = Column(Text, nullable=False)
    breed_id = Column(Integer, ForeignKey('breed.id'), nullable=False)

    # We are adding this because it's convenient, 
    # not because it's necessary:
    breed = relationship('Breed', foreign_keys='Breed.id')

    def __repr__(self):
        return '<Dog {}: {} is a {}>'.format(self.id, self.name, self.breed.type)
```

If we assume our tables are populated like the earlier example demonstrates, we can do this:
```python
>>> from models import Dog, Breed
>>> a_dog = Dog.query.get(1)
>>> a_dog.id
1
>>> a_dog.name
'Fido'
>>> a_dog.breed_id
2
>>> a_dog.breed.id
2
>>> a_dog.breed.id == a_dog.breed_id  # This will always be True
True
>>> a_dog.breed
'<Breed 2: dalmatian>'
>>> dalmatian = Breed.query.get(2)
>>> dalmatian.id
2
>>> dalmatian is a_dog.breed
True
>>> dalmatian.dogs
['<Dog 1: Fido is a dalmatian>', '<Dog 3: Rover is a dalmatian>']
```
As you can see, the FK column attribute, `Dog.breed_id`, is always an integer corresponding to a particular `Breed.id` value. The `Dog.breed` relationship is merely a shortcut for loading the associated `Breed`.

I'll get to addressing 3 and 4 later, but wanted to send these out first :)

Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: @TylerHendrickson this `Dog` to `Breed` explanation was very helpful. Basically the relationships api was built into SQLAchemy for the benefit of this direct model to model relationship is that correct? As I've been going over the bullpen codebase, specifically the models, a lot of the models have a relationship setup between other models. Is this 'optional' feature usually implemented because of its' handiness?

Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: If I'm getting too nitty gritty about the specifics about SQLAlchemy let me know, I'm unable to recognize if these details (esp. about relationships between models) are applicable to understanding how to write (better) Models and utilizing ORMs, like if this knowledge is universal or framework specific.

Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: @briana-mcn Yes, we generally implement these for all connections between models in our systems. Part of the reason for this has to do with how we've orchestrated our APIs' representations of our our model-based objects.  Even aside from the API, though, it's usually very useful to have direct access to a related object because it's cuts out the step of needing to query for it; the relationship allows us to skip one or more lines of code, essentially.

I think you're fine regarding the SQLA nitty gritty, but be aware that this knowledge is _not_ universal. Other frameworks definitely have similarities to SQLAlchemy, but the "universal" bit here has to do with the database-level concepts of relational data modelling, not the ORM-specific functionality.

File: rock_paper_scissors/api/models.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ModelBase = declarative_base()
+
+
+class Game(ModelBase):
+    __tablename__ = 'game'
+
+    id = Column(Integer, primary_key=True)
+    player_1_id = Column(Integer, ForeignKey('user.id'))
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    round_id = Column(Integer, ForeignKey('round.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Since a `Game` has multiple `Round`s, this FK should be removed (only `round.game_id` is necessary to establish this relationship).

File: rock_paper_scissors/api/models.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    rounds_choice = Column(Integer)
+    current_round = Column(Integer)
+    winner_id = Column(Integer, ForeignKey('user.id'))
+
+    player_1 = relationship('User', foreign_keys=[player_1_id])
+    player_2 = relationship('User', foreign_keys=[player_2_id])
+    winner = relationship('User', foreign_keys=[winner_id])
+    round = relationship('Round')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This should be `rounds` (plural) since it will be a list of many `Round`s

File: rock_paper_scissors/api/models.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+class Game(ModelBase):
+    __tablename__ = 'game'
+
+    id = Column(Integer, primary_key=True)
+    player_1_id = Column(Integer, ForeignKey('user.id'))
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    round_id = Column(Integer, ForeignKey('round.id'))
+    rounds_choice = Column(Integer)
+    current_round = Column(Integer)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I don't think we'll need this :)

File: rock_paper_scissors/api/models.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+    games = relationship('Game', foriegn_keys=['Game.player_1_id', 'Game.player_2_id'])
+    choices = relationship('Choice')
+
+
+class Choice(ModelBase):
+    __tablename__ = 'choice'
+
+    id = Column(Integer, primary_key=True)
+    rounds_id = Column(Integer, ForeignKey('rounds.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This should be `round_id` (singular). Also note that the table for `Round` is named `round`, so the change will need to be made in the `ForeignKey` argument as well.

File: rock_paper_scissors/api/models.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+    player_id = Column(Integer, ForeignKey('user.id'))
+    move = Column('moves', Enum('Paper', 'Rock', 'Scissors'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Should be `move` (singular), since only one value will be given to a `Choice`

File: rock_paper_scissors/api/models.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    round = relationship('Round')
+
+
+class Round(ModelBase):
+    __tablename__ = 'round'
+
+    id = Column(Integer, primary_key=True)
+    player_1_choice_id = Column(Integer, ForeignKey('choice.id'))
+    player_2_choice_id = Column(Integer, ForeignKey('choice.id'))
+    winner_id = Column(Integer, ForeignKey('user.id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I wonder if this should instead be `winning_choice_id`, and a FK to `choice.id`.

The reason for this is that we can always look to the `Choice.player_id` value to see the `User` who made the winning choice. Otherwise we'd need some logic to ensure that `winner_id` actually belongs to one of the two players for the round.

File: rock_paper_scissors/api/models.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    player_2_choice = relationship('Choice', foreign_keys=[player_2_choice_id])
+
+
+class User(ModelBase):
+    __tablename__ = 'user'
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+    score = Column(Integer)
+    choices_id = Column(Integer, ForeignKey('choice.player_id'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This FK column is invalid because it would need to point to more than one value (this is the same as my comment regarding `Game.round_id`; `Choice.player_id` establishes the relationship between `Choice` and `User` in this case.

File: rock_paper_scissors/api/models.py [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    winner_id = Column(Integer, ForeignKey('user.id'))
+    game_id = Column(Integer, ForeignKey('game.id'))
+
+    game = relationship('Game')
+    user = relationship('User')
+    player_1_choice = relationship('Choice', foreign_keys=[player_1_choice_id])
+    player_2_choice = relationship('Choice', foreign_keys=[player_2_choice_id])
+
+
+class User(ModelBase):
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I'm starting to wonder if we should call this model `Player` and the associated table `player` instead of referring to them with "user". Thoughts?

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@briana-mcn Some changes for you on the heels of our discussion at the end of last week.

**Comment by briana-mcn** [comment]
@TylerHendrickson I made some changes after our meeting on Friday evening! My understanding about models and database relationships is _much_ clearer :+1:

File: rock_paper_scissors/api/models.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
+from sqlalchemy import Column, Integer, String
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: The lines beginning with `from sqlalchemy import ` should be merged into a single import statement:
```python
from sqlalchemy import (
    Column,
    Enum,
    ForeignKey
    Integer,
    String
)
```

File: rock_paper_scissors/api/models.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+ModelBase = declarative_base()
+
+
+class Game(ModelBase):
+    __tablename__ = 'game'
+
+    id = Column(Integer, primary_key=True)
+    player_1_id = Column(Integer, ForeignKey('user.id'))
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    round_id = Column(Integer, ForeignKey('round.id'), nullable=False)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I think this [previous comment](https://github.com/briana-mcn/rock-paper-scissors/pull/25#discussion_r127747698) may have gotten lost, but it still stands :)

> Since a `Game` has multiple `Round`s, this FK should be removed (only `round.game_id` is necessary to establish this relationship).

File: rock_paper_scissors/api/models.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    rounds_remaining = relationship('Round', foreign_keys='Round.id')
+    current_round = relationship('Round', foreign_keys='Round.id')
+
+
+class Round(ModelBase):
+    __tablename__ = 'round'
+
+    id = Column(Integer, primary_key=True)
+    game_id = Column(Integer, ForeignKey('game.id'))
+    winner_id = Column(Integer, ForeignKey('user.id'), nullable=False)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Similarly, what are your thoughts on my other [previous comment](https://github.com/briana-mcn/rock-paper-scissors/pull/25#discussion_r127749639)?

> I wonder if this should instead be `winning_choice_id`, and a FK to `choice.id`.
>
> The reason for this is that we can always look to the `Choice.player_id` value to see the `User` who made the winning choice. Otherwise we'd need some logic to ensure that `winner_id` actually belongs to one of the two players for the round.

File: rock_paper_scissors/api/models.py [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+    choices = relationship('Choice', foreign_keys=['Choice.round_id'])
+
+
+class User(ModelBase):
+    __tablename__ = 'user'
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+
+    games_won = relationship('Game', foreign_keys=['Game.player_1_id', 'Game.player_2_id'])
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: The `foreign_keys` value here should point to `Game.winner_id`.

If you wanted to have a `games_played` relationship, `['Game.player_1_id', 'Game.player_2_id']` would be appropriate for something like that.

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@briana-mcn This looks _really_ good! A couple of notes for you to address.

File: rock_paper_scissors/api/models.py (hunk 26 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: After looking further into it, I agree with you. It doesn't make sense to determine the winner of each round through User.id. I think _explicit is better than implicit_ could apply here.

File: rock_paper_scissors/api/models.py (hunk 27 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: Thanks for pointing this out.

File: rock_paper_scissors/api/models.py (hunk 24 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I was wondering about that. I didn't know if imports had different structural purposes in a program if they needed to be imported differently...typing this out made me realize that this is not the case, as the interpreter wouldn't need that specification :smiley:

File: rock_paper_scissors/api/models.py (hunk 25 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: :+1: Got it, makes sense after thinking it through

File: rock_paper_scissors/api/models.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class User(ModelBase):
+    __tablename__ = 'user'
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+
+    games_won = relationship('Game', foreign_keys=['Game.winner_id'])
+    rounds_won = relationship('Game', foreign_keys=['Round.winner_id'])
```
Comment (briana-mcn) [inline rock_paper_scissors/api/models.py]: @TylerHendrickson I haven't made changes to this because I was unsure of the foreign_keys value.
my gut feeling would be `games_won = relationship('Round', foreign_keys=['Round.winning_choice_id'])`
I hesitate because the relationship wouldn't explicitly reference the winner with the User.id

File: rock_paper_scissors/api/models.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+    rounds_remaining = relationship('Round', foreign_keys='Round.id')
+    current_round = relationship('Round', foreign_keys='Round.id')
+
+
+class Round(ModelBase):
+    __tablename__ = 'round'
+
+    id = Column(Integer, primary_key=True)
+    game_id = Column(Integer, ForeignKey('game.id'))
+    winning_choice_id = Column(Integer, ForeignKey('choice.id'), nullable=False)
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: This will need to be made nullable, since we won't know the winning choice at the time that a round is initially saved to the database.

Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: Additionally, let's add another `relationship` attribute for `winning_choice`

File: rock_paper_scissors/api/models.py [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    rounds_limit = Column(Integer)
+    winner_id = Column(Integer, ForeignKey('user.id'), nullable=False)
+
+    player_1 = relationship('User', foreign_keys=[player_1_id])
+    player_2 = relationship('User', foreign_keys=[player_2_id])
+    winner = relationship('User', foreign_keys=[winner_id])
+
+    rounds_remaining = relationship('Round', foreign_keys='Round.id')
+    current_round = relationship('Round', foreign_keys='Round.id')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I think this should be an attribute using `@property` instead of a `relationship`. It should return the associated `Round` which is currently incomplete, or `None` if all are complete.

File: rock_paper_scissors/api/models.py [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+    player_1_id = Column(Integer, ForeignKey('user.id'))
+    player_2_id = Column(Integer, ForeignKey('user.id'))
+    rounds_limit = Column(Integer)
+    winner_id = Column(Integer, ForeignKey('user.id'), nullable=False)
+
+    player_1 = relationship('User', foreign_keys=[player_1_id])
+    player_2 = relationship('User', foreign_keys=[player_2_id])
+    winner = relationship('User', foreign_keys=[winner_id])
+
+    rounds_remaining = relationship('Round', foreign_keys='Round.id')
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: I think this should be an attribute using `@property` that returns an integer representing the _number_ of rounds which are incomplete, based on `rounds_limit`. Basically:
```python
@property
def rounds_remaining(self):
    rounds_complete = 0

    for round in self.rounds:
        if round.winning_choice is not None:
            rounds_complete += 1

    return self.rounds_limit - rounds_complete
```

Note that we'll also need a `rounds` relationship based on the `Round.game_id` FK.

File: rock_paper_scissors/api/models.py (hunk 28 above)
Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: You're correct on this. As discussed, we should instead make a `rounds_won` attribute (using `@property`). We can't base it on an iteration of associated `games_one`, however, because a player may have won rounds for a game which is currently incomplete.

Instead, we'll need to query for the data!

Specifically, we need all `Round`s where the `winning_choice` is associated with the user. When you get here, let me know and we can work on the query together.

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@briana-mcn A few notes for you to address. Additionally, now would be a good time to look through all your FK columns and ensure that a corresponding `relationship` exists for each one.

File: rock_paper_scissors/api/models.py (hunk 29 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: I was under the impression that when saving an instance to the database with a field that isn't populated but will be present _eventually_ that `NOT NULL` is used? I wonder where I've mixed this up.

Comment (TylerHendrickson) [reply rock_paper_scissors/api/models.py]: The terminology here is what's confusing things, I think :)

`NULL` is a value (of sorts) which represents "no value". If a column is NULLable, it means it's allowed to hold cells which have no value- NULL. Contrastingly, if a column is NOT NULLable, no cells in that column are allowed to be NULL (i.e. everything must have a value, and that value cannot be `NULL`).

To break it down more simply:
- If a column should be able to contain `NULL` values, then the column is "NULL-able", so `nullable=True`.
- If a column should _not_ be able to contain `NULL` values, then the column is "NOT NULL-able", so `nullable=False`.

File: rock_paper_scissors/api/models.py (hunk 31 above)
Comment (briana-mcn) [reply rock_paper_scissors/api/models.py]: Thanks for the breakdown! This is clearer to me. I  thought that models were much different from classes- not thinking that methods were available to models... ...

File: rock_paper_scissors/api/models.py [hunk 32]
```diff
@@ -19,7 +19,7 @@ class Game(ModelBase):
     player_1_id = Column(Integer, ForeignKey('user.id'))
     player_2_id = Column(Integer, ForeignKey('user.id'))
     rounds_limit = Column(Integer)
-    winner_id = Column(Integer, ForeignKey('user.id'), nullable=False)
+    winner_id = Column(Integer, ForeignKey('user.id'), nullable=True)
```
Comment (briana-mcn) [inline rock_paper_scissors/api/models.py]: Will I want to create `nullable` parameters in the other Columns of the models for the module that don't have ForeignKeys?

File: rock_paper_scissors/api/models.py [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+        return rounds
+
+
+class Choice(ModelBase):
+    __tablename__ = 'choice'
+
+    id = Column(Integer, primary_key=True)
+    round_id = Column(Integer, ForeignKey('round.id'))
+    player_id = Column(Integer, ForeignKey('user.id'))
+    move = Column('moves', Enum('Paper', 'Rock', 'Scissors'))
```
Comment (TylerHendrickson) [inline rock_paper_scissors/api/models.py]: Let's set the column name to `move` as well to stay consistent with the attribute name.
